dev-docs/RFCs/v6.1/transform-webgl1-support.md

```diff
@@ -39,17 +40,195 @@ We can start implementation with Option#1.
 
 Once the render target is setup, and data is rendered we can use `gl.readPixels` to read the data.
 
-### Inputs and Outputs
+### Inputs
+
+Under WebGL2, all inputs are luma.gl `Buffer` objects. Due to WebGL1 limitation of buffer API (no async Texture to Buffer API) supporting `Buffer` objects require a CPU GPU sync for writing results from texture (render target) into a `Buffer`  object. To avoid such a sync and able to run a sequence of Transform operations (run -> swap -> run ...) we need to support luma.gl `Texture` objects as inputs and outputs.
+
+Inputs can be specified in following three formats :
+
+1. TypedArray (WebGL1 and WebGL2, ex: Float32Array, Uint8Array, etc)
+
+When using WebGL1, internally `Texture` object is built with user provided `TypedArray` and used to provided input to the VertexShader.
+
+2. Texture2D (WebGL1 and WebGL2)
+
+Supported by both WebGL1 and WebGL2, but only the formats that are supported by WebGL context being used.
+
+3. Buffer (WebGL2, luma.gl `Buffer` object)
+
+Only supported in WebGL2 and set as `attributes` on the `VertexShader`.
+
+
+* It is also possible to use any combination of above input methods when using WebGL2 and any combination of #1 and #2 when using WebGL1.
+
+To achieve this flexibility a minor change is needed in how VertexShader source (`vs` parameter) is specified on `Tranform` object. In existing version `vs` needs to specify all inputs using `attributes` (GLSL100) or `in` (GLSL300), that should be now removed as explained in below sample code
+
+#### Sample code
+
+##### Current usage
+
+```
+const VS = `\
+attribute float leftValue;
+attribute float rightValue;
+varying float outValue;
+
+void main()
+{
+  outValue = leftValue + rightValue;
+}
+`;
+
+...
+
+const leftValueBuffer = new Buffer(gl, { ... });
+const rightValueBuffer = new Buffer(gl, { ... });
+const outValueBuffer = new Buffer(gl, { ... });
+
+const transform = new Transform(gl, {
+  sourceBuffers: {
+    leftValue: leftValueBuffer,
+    rightValue: rightValueBuffer,
+  },
+  vs: VS,
+  feedBackBuffers: {
+    outValue: outValueBuffer
+  },
+  varyings: ['outValue'],
+  elementCount: 5
+});
+```
+
+##### Proposed usage
+
+VertexShader source shouldn't specify `attribute`/`in` input declaration.
```

Comment:
Would probably require a little more work, but my preference would that the we keep writing shaders for the WebGL2 case, and automatically strip out and replace the attributes when in WebGL1.

Long-term there is basically a choice between creating our own abstract shader language that we can generate lower level shaders from. A number of graphics frameworks does this. But I would prefer that we let the user work with standard GLSL as much as possible.

That said, we can absolutely prototype this per your suggestion, and add the transpilation later.